tests/filecheck/dialects/csl/csl-stencil.md

```diff
@@ -0,0 +1,290 @@
+# Stencil to CSL lowering
+
+Proposal: A `csl_stencil` dialect that manages communicating data across
+the stencil pattern, sending to plus receiving from all neighbours.
+
+Starting point: Output of the `stencil-tensorize-z-dim` transform. This representation
+expresses that each PE holds one vector of z-values, while accesses to differing x/y coordinates
+require communicating with neighbouring PEs.
+
+The goal is to more closely express on a high level what our low-level stencil (communications)
+library does.
+
+## Starting point 
+Recall the following starting representation of tensorized z-dim values - please note, the
+`tensor.extract_slice` ops manage the halo and have been slightly re-ordered for improved readability: 
+
+``` mlir
+builtin.module {
+  func.func @gauss_seidel(%a : memref<1024x512xtensor<512xf32>>, %b : memref<1024x512xtensor<512xf32>>) {
+    %0 = stencil.external_load %a : memref<1024x512xtensor<512xf32>> -> !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>>
+    %1 = stencil.load %0 : !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>> -> !stencil.temp<[-1,1023]x[-1,511]xtensor<512xf32>>
+    %2 = stencil.external_load %b : memref<1024x512xtensor<512xf32>> -> !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>>
```

Comment:
Unless I missed something specific to the tensorized aspect of it, you could write the function as taking the `!stencil.field`s directly here, they would get lowered to the memref you are converting from anyway :slightly_smiling_face:

Reply:
Always happy to work with as many kernels as I can get my hands on and any variety of representations we might want to support (this particular one was provided by @mesham, with the tensorisation pass applied to it).